### PR TITLE
py/objdict: Disallow possible modifications to fixed dicts.

### PR DESCRIPTION
**Pull request by mikewadsten**
Example of bad code when running a non-debug Unix build:

```python
>>> import uerrno
>>> uerrno.errorcode.setdefault(7777, 1)
zsh: segmentation fault (core dumped)  ./micropython
```

`mp_map_lookup` does assert if it is called with a fixed map and lookup_kind other than MP_MAP_LOOKUP, but asserts don't stop bad code paths from executing when asserts are disabled (i.e. "release" builds).

On an embedded system, the `setdefault` call above does not cause any crashes, and in fact will eventually lead to a MemoryError when we detect that the map in question is not in the GC pool, but not before updating `map->alloc` (see py/map.c line 194). On a Unix build, this line leads to a segmentation fault.

`uerrno.errorcode` was the only obvious "fixed" dictionary I could find in the MicroPython core and mainline ports that is directly user-accessible (all others are things like the globals and locals tables for various modules), but any port could conceivably add their own constant dict to a module's table, where `setdefault` would lead to this same bad behavior.

### CODE REVIEW

**Comment by pfalcon** [comment]
This should be RuntimeError.

**Comment by mikewadsten** [comment]
> This should be RuntimeError.

Okay, I can make that change. Should I keep the error message as is, or pass in NULL (so no message) to try to save some code/data space assuming this error shouldn't be hit under normal circumstances?

**Comment by pfalcon** [comment]
> pass in NULL (so no message)

My opinion would be that, yeah.

**Comment by mikewadsten** [comment]
Okay, done. I have also updated the errno1.py test to check this case (so that coverage doesn't drop).

**Comment by mikewadsten** [comment]
At first I thought the fact that Travis hasn't built yet was my fault (maybe because of my force-push), but nope. https://www.traviscistatus.com/incidents/c1s4dlyxd4lf

**Comment by pfalcon** [comment]
@mikewadsten : Thanks for updates, +1 from me.

**Comment by mikewadsten** [comment]
Since I can't merge this (due to the code size increase), what's the next step? Get @dpgeorge to look it over and merge manually?

**Comment by dhylands** [comment]
You can't merge it anyways. I believe that @dpgeorge is currently the only one with write permissions on the repository, so all changes need to go through him.

**Comment by mikewadsten** [comment]
@dhylands That's a fair point. I suppose my previous comment can be ignored then :)

**Comment by dpgeorge** [comment]
@mikewadsten thanks for raising this issue, and providing a fix.

But it requires a bit more thought because there are at least 6 independent cases of this (that I could find).  The following all make the assertion at the start of mp_map_lookup() fail:
```python
int.from_bytes = 1
del int.from_bytes
uerrno.errorcode.clear()
uerrno.errorcode.popitem()
uerrno.errorcode.pop(0)
uerrno.errorcode.setdefault(0, 0) # same path as .pop()
uerrno.errorcode.update([(1,2)]) # there are 3 variants of this depending on the args
```
And in CPython the first two raise TypeError, so probably the other cases should as well (because a fixed/frozen dict is not the correct type to act upon with these mutating methods).

**Comment by mikewadsten** [comment]
@dpgeorge I agree that raising TypeError when making a call that might modify a fixed/frozen dict seems like the right thing to do here. If clear and popitem called mp_map_lookup, then I think this fix would have covered all the cases (because dict_get_helper will raise a KeyError or silently skip the operation for everything but MP_MAP_LOOKUP_ADD_IF_NOT_FOUND).

Perhaps there exists minimal way to add a helper function like `void mp_ensure_not_fixed(const mp_obj_dict_t *);` to raise TypeError if the dict is fixed?

**Comment by dpgeorge** [comment]
In b45c8c17f0f295e919a90659d4c1880a47b228c1 I fixed the case for store/delete on a built-in type (eg `int.to_bytes=1` now raises an exception).  This was independent (and the fix was independent) to the issues with dict.

> If clear and popitem called mp_map_lookup, then I think this fix would have covered all the cases

Yes it would have, but clear and popitem are distinctly different in their implementation so need a separate check.

One option is to create a new type for frozen dicts, and not include any of the mutating methods in its list of methods, but it looks like that will add a lot of bytes to the code.

> Perhaps there exists minimal way to add a helper function like void mp_ensure_not_fixed(const mp_obj_dict_t *); to raise TypeError if the dict is fixed?

Yes that looks like the best option, to fix all the dict cases (which look to be the only remaining ones).  This function is so simple that it could probably be a macro/inline function and not increase code, or even just write it out explicitly 3 times because clear and popitem don't need to check the lookup_kind variable.

@mikewadsten can you have a go at making the change, and also adding tests for all the cases in dict that previously failed?

**Comment by mikewadsten** [comment]
@dpgeorge It looks like the change introduces 12 bytes of code in each site.

Build environment:

  * `arm-none-eabi-gcc` version 7.2.1 (using latest `ryankurte/docker-arm-embedded` Docker image)
  * command line: `make -C /src/ports/minimal CROSS=1 build/firmware.bin`

Change:

```c
if (self->map.is_fixed) {
    mp_raise_TypeError(NULL);
}
```

Naturally the size change grows if I pass a non-NULL message, such as "built-in dict".

Do you have any ideas on ways to minimize the code size impact (i.e. can you maybe try on a different compiler version), or do you think this level of impact is unavoidable?

**Comment by dpgeorge** [comment]
> Do you have any ideas on ways to minimize the code size impact (i.e. can you maybe try on a different compiler version), or do you think this level of impact is unavoidable?

It's pretty unavoidable.  The only think you can try is making it a function that is reused, but it probably costs about 8 bytes just to call the function so would increase total size by more than just writing it inline.  If you can report the total code size increase for "all inline" vs "all calling a common function" that would be useful.

**Comment by mikewadsten** [comment]
Here's the results on my machine:

|| baseline (bbb0843) | all inline | common function
------|--------------------|------------|---------------
minimal (gcc 7.2.1) | 71,008 | 71,048 (+40) | 71,048 (+40)
unix x64 (gcc 4.8.2) | 380,044 | 380,092 (+48) | 380,156 (+112)

I get the same results if I roll back 24 hours (to 771dfb0) - both versions add 48 bytes over the baseline minimal size.

I could swear when I tried it out last night, the common-function version was at least a little bigger... meh.

**Comment by dpgeorge** [comment]
@mikewadsten thanks for providing the metrics.  It looks like all-inline is the way to go (although it depends if the changes you report there fix all cases or not...).

> these changes (clear, popitem, dict_get_helper) doesn't fix update or uerrno.errorcode[1] = 'foo'.

It would be good if you could push all the required changes to fix all the cases.  It'll require separate checks in the various functions.

**Comment by mikewadsten** [comment]
I'm working on wrapping up these fixes right now. Looks like the only thing I have left to clear up in the `uerrno.errorcode[1] = 'foo'` case. (I'll check `del` while I'm at it, since if I remember correctly, that goes through the same code path?)

**Comment by mikewadsten** [comment]
Here are my results for today:

|| baseline (bbb0843) | all inline | common function
------|--------------------|------------|---------------
minimal (gcc 7.2.1) | 71,000 | 71,072 (+72) | 71,064 (+64)
pybv11 (gcc 7.2.1) | 324,484 | 324,560 (+76) | 324,552 (+68)
unix x64 (gcc 4.8.2) | 380,044 | 380,124 (+80) | 380,188 (+144)

(These are all the 'text' counts that get reported after the relevant `make` command. The deltas are based on applying the fix to bbb0843. I don't know why the minimal build dropped 8 bytes from last week, when I'm building the same code on the same machine in the same Docker image...)

So basically it now looks like the cost of the common function call amortizes to less code on the embedded targets, but increasingly more on the Unix target. I tried marking it MP_NOINLINE but that had no effect on the size.

@dpgeorge Given the new results, should I move forward with a common function (`STATIC void mp_ensure_not_fixed(const mp_obj_dict_t *dict)`)?

**Comment by dpgeorge** [comment]
> Given the new results, should I move forward with a common function (STATIC void mp_ensure_not_fixed(const mp_obj_dict_t *dict))?

Yes that's a good way forward.

File: tests/basics/fixed_dict.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    raise SystemExit
+
+# Save a copy of uerrno.errorcode, so we can check later
+# that it hasn't been modified.
+errorcode_copy = uerrno.errorcode.copy()
+
+try:
+    uerrno.errorcode.popitem()
+except TypeError:
+    pass
```
Comment (dpgeorge) [inline tests/basics/fixed_dict.py]: Please make this a `print('TypeError')` because otherwise there's no way to tell if an exception was raised or not, and in this case an exception should be raised.  Then you'll need to make a .exp file that has the expected output (because this script can't be checked by CPython).

**Comment by dpgeorge** [comment]
The test name should probably be called `dict_fixed.py` so it has the same prefix as other dict tests.

**Comment by mikewadsten** [comment]
Amended and force-pushed with requested changes.

**Comment by dpgeorge** [comment]
Thanks @mikewadsten!
